src/editorwidget/widget.ts

```diff
@@ -61,52 +65,33 @@ class EditorWidget extends CodeEditorWidget {
   /**
    * Construct a new editor widget.
    */
-  constructor(
-    editorFactory: (host: Widget) => CodeEditor.IEditor,
-    context: DocumentRegistry.Context,
-    editorMimeTypeService: IEditorMimeTypeService) {
-    super(editorFactory);
+  constructor(options: EditorWidget.IOptions) {
+    super(options.factory);
     this.addClass(EDITOR_CLASS);
-    this._context = context;
+    let context = this._context = options.context;
+    this._mimeTypeService = options.mimeTypeService;
+
     let model = context.model;
     let editor = this.editor;
     let value = editor.model.value;
 
     // Prevent the initial loading from disk from being in the editor history.
     context.ready.then( () => {
-      value.text = model.toString();
-      editor.model.clearHistory();
+      if (!this.isDisposed) {
+        value.text = model.toString();
+        editor.model.clearHistory();
+      }
     });
 
     value.text = model.toString();
-
     this.title.label = context.path.split('/').pop();
-    model.stateChanged.connect((m, args) => {
-      if (args.name === 'dirty') {
-        if (args.newValue) {
-          this.title.className += ` ${DIRTY_CLASS}`;
-        } else {
-          this.title.className = this.title.className.replace(DIRTY_CLASS, '');
-        }
-      }
-    });
-    model.contentChanged.connect(() => {
-      let old = value.text;
-      let text = model.toString();
-      if (old !== text) {
-        value.text = text;
-      }
-    });
-    this.editor.model.value.changed.connect((sender, args) => {
-      model.fromString(value.text);
-    });
-    editor.model.mimeType = editorMimeTypeService.getMimeTypeByFilePath(context.path);
-    context.pathChanged.connect((c, path) => {
-      editor.model.mimeType = editorMimeTypeService.getMimeTypeByFilePath(path);
-      this.title.label = path.split('/').pop();
-    });
+    this._handleDirtyState();
+    this._onPathChanged();
 
-    // TODO disconnect on deactivation
+    model.stateChanged.connect(this._onModelStateChanged, this);
+    model.contentChanged.connect(this._onContentChanged, this);
```

Comment:
Should not be they registered only on `context.ready`? Otherwise `this._onContentChanged` can be called before the context is ready.